kafka/socks5_wrapper.py

```diff
@@ -0,0 +1,236 @@
+            data = self._sock.recv(bytes_remaining)
+            if not data:
+                break
+            self._buffer_in = self._buffer_in + data
+
+        return self._buffer_in
+
+    def _read_buf(self, datalen):
+        """Read and consume bytes from socket connection
+
+        It's expected that the caller handles e.g. blocking exceptions"""
+        buf = self._peek_buf(datalen)
+        if buf:
+            self._buffer_in = self._buffer_in[len(buf):]
+        return buf
+
+    def connect_ex(self, addr):
+        """Runs a state machine through connection to authentication to
+        proxy connection request.
+
+        The somewhat strange setup is to facilitate non-intrusive use from
+        BrokerConnection state machine."""
+
+        if self._state == ProxyConnectionStates.DISCONNECTED:
+            self._state = ProxyConnectionStates.CONNECTING
+
+        if self._state == ProxyConnectionStates.CONNECTING:
+            _, _, _, _, sockaddr = self._proxy_addr
+            ret = self._sock.connect_ex(sockaddr)
+            if not ret or ret == errno.EISCONN:
+                self._state = ProxyConnectionStates.NEGOTIATE_PROPOSE
+            else:
+                return ret
+
+        if self._state == ProxyConnectionStates.NEGOTIATE_PROPOSE:
+            if self._proxy_url.username and self._proxy_url.password:
+                # Propose username/password
+                self._buffer_out = b"\x05\x01\x02"
+            else:
+                # Propose no auth
+                self._buffer_out = b"\x05\x01\x00"
+            self._state = ProxyConnectionStates.NEGOTIATING
+
+        if self._state == ProxyConnectionStates.NEGOTIATING:
+            if self._flush_buf() is True:
+                buf = self._read_buf(2)
+                if buf[0:1] != b"\x05":
+                    log.error("Unrecognized SOCKS version")
+                    self._state = ProxyConnectionStates.DISCONNECTED
+                    self._sock.close()
+                    return errno.ECONNREFUSED
+
+                if buf[1:2] == b"\x00":
+                    # No authentication required
+                    self._state = ProxyConnectionStates.REQUEST_SUBMIT
+                elif buf[1:2] == b"\x02":
+                    userlen = len(self._proxy_url.username)
+                    passlen = len(self._proxy_url.password)
+                    self._buffer_out = struct.pack(
+                        "!bb{}sb{}s".format(userlen, passlen),
+                        1,
+                        userlen,
+                        self._proxy_url.username.encode(),
+                        passlen,
+                        self._proxy_url.password.encode(),
+                    )
+                    self._state = ProxyConnectionStates.AUTHENTICATING
+
+        if self._state == ProxyConnectionStates.AUTHENTICATING:
+            if self._flush_buf() is True:
+                buf = self._read_buf(2)
+                if buf[0:2] == b"\x01\x00":
```

Comment:
Only requested two bytes so `[0:2]` doesn't do anything and just looks a bit confusing, same for other similar places.

Reply:
Adjusted. This was a bit of a leftover decoding one byte at a time - version + return code.